File: src/main/java/cs/java/collections/CSListImpl.java
package cs.java.collections;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

import static cs.java.lang.CSLang.empty;
import static java.util.Arrays.asList;

public class CSListImpl<T> extends ArrayList<T> implements CSList<T> {

    public CSListImpl() {
        super();
    }

    public CSListImpl(Collection<? extends T> collection) {
        super(collection);
    }

    public CSListImpl(int capacity) {
        super(capacity);
    }

    public CSListImpl(java.util.List<T> list) {
        super(list);
    }

    public T at(int index) {
        if (index < size() && index >= 0) return get(index);
        return null;
    }

    public int delete(T item) {
        int index = indexOf(item);
        remove(index);
        return index;
    }

    public T first() {
        return at(0);
    }

    public T second() {
        return at(1);
    }

    public int index(T item) {
        return indexOf(item);
    }

    public boolean getHasItems() {
        return size() > 0;
    }

    public boolean isLast(T item) {
        return last() == item;
    }

    public T last() {
        return at(lastIndex());
    }

    public T previous() {
        return at(lastIndex() - 1);
    }

    public int lastIndex() {
        return size() - 1;
    }

    public CSList<T> range(int fromIndex) {
        return range(fromIndex, size());
    }

    public CSList<T> range(int fromIndex, int toIndex) {
        return new CSListImpl<T>(subList(fromIndex, toIndex));
    }

    public T removeLast() {
        if (isEmpty()) return null;
        return remove(size() - 1);
    }

    public int count() {
        return size();
    }

    public boolean has(T object) {
        return super.contains(object);
    }

    public int length() {
        return size();
    }

    public CSList<T> reverse() {
        Collections.reverse(this);
        return this;
    }

    public CSList<T> removeAll() {
        clear();
        return this;
    }

    public T put(T item) {
        add(item);
        return item;
    }

    public T add(T item, int index) {
        add(index, item);
        return item;
    }

    public T set(T item, int index) {
        set(index, item);
        return item;
    }

    public CSList<T> append(T... items) {
        if (empty(items)) return this;
        addAll(asList(items));
        return this;
    }

    public CSList<T> append(List<T> items) {
        if (empty(items)) return this;
        addAll(items);
        return this;
    }

    public CSList<T> insert(int index, T item) {
        add(index, item);
        return this;
    }

    public CSList<T> reload(List<T> values) {
        clear();
        addAll(values);
        return this;
    }

    public CSList<T> values() {
        return this;
    }
}

File: library/src/main/java/renetik/java/collections/CSListImpl.kt
package renetik.java.collections

import java.util.*
import java.util.Arrays.asList

class CSListImpl<T> : ArrayList<T>, CSList<T> {

    override val hasItems get() = size > 0

    override val values get() = this

    constructor() : super() {}
    constructor(collection: Collection<T>) : super(collection) {}
    constructor(capacity: Int) : super(capacity) {}
    constructor(list: List<T>) : super(list) {}

    override fun at(index: Int): T? = if (index < size && index >= 0) get(index) else null
    override fun first(): T? = at(0)
    override fun second(): T? = at(1)
    override fun isLast(item: T) = last() === item
    override fun isLastIndex(index: Int) = index == size - 1
    override fun last() = at(lastIndex())
    override fun previousLast() = at(lastIndex() - 1)
    override fun lastIndex() = size - 1
    override fun rangeFrom(fromIndex: Int) = range(fromIndex, size)
    override fun range(fromIndex: Int, toIndex: Int) = CSListImpl(subList(fromIndex, toIndex))
    override fun deleteLast() = if (isEmpty()) null else removeAt(size - 1)
    override fun reverse() = apply { Collections.reverse(this) }

    override fun put(item: T): T {
        add(item)
        return item
    }

    override fun putAll(items: Iterable<T>) = apply { for (item in items) add(item) }
    override fun putAll(vararg items: T) = putAll(asList(*items))

    override fun put(item: T, index: Int): T {
        add(index, item)
        return item
    }

    override fun replace(item: T, index: Int): T {
        set(index, item)
        return item
    }

    override fun reload(values: Iterable<T>) = apply {
        clear()
        putAll(values)
    }

    override fun delete(item: T): Int {
        val index = indexOf(item)
        removeAt(index)
        return index
    }

    override fun delete(index: Int): T? = at(index)?.apply { super.removeAt(index) }

    override fun deleteAll() = apply { clear() }
}
